### ISSUE TITLE
BLAKE3

### ISSUE THREAD

**Issue by sergeevabc** [rhash/RHash#157]
Dear @rhash,

BLAKE3 seems to be capable to replace the whole zoo of MD5/SHA1/2/3/Blake2 flavors to check files for malicious changes, while xxHash 128 is a relevant CRC32 replacement for catching accidental changes (e.g. network transmission errors and data rot). This [consensus is maturing][1] as we speak. (cc: @Cyan4973)

[1]: https://github.com/Cyan4973/xxHash/issues/479#issuecomment-801528938

Alas, `b3sum`, the official CLI implementation of BLAKE3, has a [serious drawback][2] related to memory management and its devs are far from fixing it (cc: @veorq). So until now I’ve been fortunate to use the advantages of this algorithm mainly thanks to GUI apps such as [RapidCRC][3] by @ov2 and [HashCheck][4] fork by @idrassi. I believe a proper CLI implementation would benefit the community.

Be kind to add BLAKE3 (and sooner or later [xxHash][5] 128) to RHash, please.

[2]: https://github.com/BLAKE3-team/BLAKE3/issues/78#issuecomment-801425510
[3]: https://www.ov2.eu/programs/rapidcrc-unicode/
[4]: https://github.com/idrassi/HashCheck/releases/ 
[5]: https://github.com/rhash/RHash/issues/134

**Comment by data-man** [comment]
Non-optimized and public domain https://github.com/michaelforney/blake3-tiny

**Comment by PapaKarl0** [comment]
@sergeevabc 
There's a third GUI implementation - TurboSFV (which, incidentally, also has a neat CLI version). It's pretty fast, I think the fastest out of three.

**Comment by PapaKarl0** [comment]
Btw, I have done some tests, and according to my results TurboSFV cli version is x2 faster than RapidCRC, and about 40% faster than HashCheck. Not trying to promote turbo software, which is not free.

**Comment by QuAzI** [comment]
Still looks like Blake3 most perspective and fast to check files

**Comment by AxelPetermann** [comment]
Also vote for blake3 implementation

**Comment by sergeevabc** [comment]
Err… Aleksey Kravchenko, hello?

**Comment by rhash** [comment]
> Non-optimized and public domain https://github.com/michaelforney/blake3-tiny

Will community benefit from a non-optimized Blake3?
This way we lose the strongest point – the speed of the algorithm.

Blake3 optimizations usualy include SSE4.1, AVX2, AVX512 and other instruction sets. It's will take time to support them.

**Comment by sergeevabc** [comment]
@rhash, aren't you overthinking here?

My laptop, for example, has Core2Duo CPU with SSE3 , no SSE4 or AVX is available. And home servers with backups and torrents, where files are hashed from time to time to detect data degradation, are built with HDDs -- they are naturally slower to read files than the most non-optimized SIMD version of Blake3. Benchmarks like [Kopia](https://github.com/kopia/kopia/issues/4181)'s one show the advantage of Blake3 over SHA256, but my setup cannot utilize all its power. The point, Aleksey, is to use the latest achievement of mathematical thought with a proven high margin of safety.

Blake3 implementations I'm more or less happy with
* GUI: [CHK](https://compressme.net/), [RapidCRCUnicode](https://www.ov2.eu/programs/rapidcrc-unicode), [TurboSFV](https://turbosfv.com/)
* CLI: Andrew Baker [filehash](https://brainwavecc.com/brainwave-utilities/), Dmitry Dulesov [hashsum](https://github.com/ddulesov/hashsum), Tino Reichardt [7zipZSTD](https://github.com/mcmilk/7-Zip-zstd/blob/master/C/hashes/blake3.c)

**Comment by rhash** [comment]
Blake3 hash function is supported without hardware optiomizations by 1df9e4de40a1d99271547f3f851d170d0725e27c, a384421cc718940fcce84a0ba6014aa34467c0bd.

Blake3 speed comparison on Intel Core i5:
| # | Hash function|Speed (MBps)|  CPB   |
|---|--------------|------------|--------|
| 1 | CRC32        | 1700,00    |  0,78  | 
| 2 | SHA1         | 732,47     |  1,88  |
| 3 | MD5          | 868,90     |  1,63  |
| 4 | EDON-R512    | 600,23     |  2,16  |
| 5 | **BLAKE3**   | **270,97** |**4,72**|
| 6 | BLAKE2B      | 235,51     |  5,71  |

**Comment by rhash** [comment]
**BLAKE3** is now part of [RHash v1.4.6](https://github.com/rhash/RHash/releases/tag/v1.4.6).
